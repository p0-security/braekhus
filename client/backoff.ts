import { createLogger } from "../log/index";
import { deferral } from "../util/deferral";

const logger = createLogger({ name: "backoff" });

export class Backoff {
  #startMillis;
  #maxMillis;
  #count = 0;

  constructor(startMillis: number, maxMillis: number) {
    if (startMillis <= 0) {
      throw new Error("startMillis must be greater than 0");
    }
    if (maxMillis < startMillis) {
      throw new Error("maxMillis must be greater than or equal to startMillis");
    }
    this.#startMillis = startMillis;
    this.#maxMillis = maxMillis;
  }

  next() {
    this.#count++;
    return Math.min(
      this.#startMillis * Math.pow(2, this.#count - 1),
      this.#maxMillis
    );
  }

  reset() {
    this.#count = 0;
  }
}

export type RetryOptions = {
  startMillis: number;
  maxMillis: number;
  maxRetries: number;
};

export const sleep = (millis: number): Promise<void> => {
  return new Promise<void>((resolve) => {
    setTimeout(resolve, millis);
  });
};

export const retryWithBackoff = async <T>(
  options: RetryOptions,
  func: () => Promise<T>
) => {
  const result = deferral<T>();

  const run = async () => {
    const { startMillis, maxMillis, maxRetries } = options;
    const backoff = new Backoff(startMillis, maxMillis);
    let count = 0;
    while (true) {
      try {
        result.resolve(await func());
        return;
      } catch (e: any) {
        count++;
        if (count > maxRetries) {
          result.reject(e);
          return;
        }
        const timeout = backoff.next();
        logger.debug(
          { nextTimeout: timeout, count, error: e },
          "Retrying with backoff timeout"
        );
        await sleep(timeout);
      }
    }
  };

  run();

  return result.promise;
};
