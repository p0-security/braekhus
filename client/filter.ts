import { isArray } from "lodash";

import { createLogger } from "../log";

const logger = createLogger({ name: "filter" });

// import { jq } from "node-jq";
// TODO fix node config because jq-node cannot be imported with the `import` syntax above
// Error:
// Module '"node-jq"' has no exported member 'jq'.
const jq = require("node-jq");

export const jqTransform = async (
  data: any,
  jqHeader: string | string[] | undefined
): Promise<any> => {
  const query = isArray(jqHeader) ? jqHeader[0] : jqHeader;
  if (!query) {
    return data;
  }
  try {
    return await jq.run(query, data, { input: "json", output: "json" });
  } catch (error: any) {
    logger.error(
      { error, jpSelectQuery: query },
      "Error running jq query, ignoring filters"
    );
    return data;
  }
};
