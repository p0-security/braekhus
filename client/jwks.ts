import { privateKeyFile, publicKeyFile } from "../util/jwk-file";
import * as jose from "jose";
import * as fs from "node:fs/promises";
import pinoLogger from "pino";

const ALG = "ES384"; // Elliptic curve with 384-bit SHA

const logger = pinoLogger({ name: "jwks" });

const loadKey = async (path: string) => {
  try {
    const data = await fs.readFile(privateKeyFile(path), {
      encoding: "utf-8",
    });
    const key = jose.importJWK(JSON.parse(data));
    return key;
  } catch (error: any) {
    logger.warn(error);
    return undefined;
  }
};

const generateKey = async (path: string) => {
  const { publicKey, privateKey } = await jose.generateKeyPair(ALG);
  await fs.writeFile(
    privateKeyFile(path),
    JSON.stringify(await jose.exportJWK(privateKey), undefined, 2),
    {
      encoding: "utf-8",
    }
  );
  await fs.chmod(privateKeyFile(path), "400");
  await fs.writeFile(
    publicKeyFile(path),
    JSON.stringify(await jose.exportJWK(publicKey), undefined, 2),
    {
      encoding: "utf-8",
    }
  );
  return privateKey;
};

const ensureKey = async (path: string) => {
  const existing = await loadKey(path);
  if (existing) return existing;
  return await generateKey(path);
};

export const jwt = async (path: string, clientId: string) => {
  const key = await ensureKey(path);
  const jwt = await new jose.SignJWT({ "tunnel-id": "my-tunnel-id" })
    .setProtectedHeader({ alg: ALG })
    .setIssuedAt(Math.floor(Date.now() * 1e-3))
    .setExpirationTime(Math.floor(Date.now() * 1e-3 + 7 * 86400)) // One week
    .setAudience("p0.dev") // TODO: Make environment-specific
    .setSubject(clientId)
    .setIssuer("kd-client")
    .sign(key);

  return jwt;
};
