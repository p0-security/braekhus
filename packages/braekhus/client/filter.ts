import { isArray } from "lodash-es";
import { run as jq } from "node-jq";

import { createLogger } from "../log/index.js";

const logger = createLogger({ name: "filter" });

export const jqTransform = async (
  data: any,
  jqHeader: string | string[] | undefined
): Promise<any> => {
  const query = isArray(jqHeader) ? jqHeader[0] : jqHeader;
  if (!query) {
    return data;
  }
  try {
    return await jq(query, data, { input: "json", output: "json" });
  } catch (error: any) {
    logger.error(
      { error, jpSelectQuery: query },
      "Error running jq query, ignoring filters"
    );
    return data;
  }
};
