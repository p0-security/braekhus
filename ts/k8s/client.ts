import {
  Create,
  Delete,
  IndirectK8sApiResponse,
  Patch,
  Read,
  Replace,
} from "../types";
import {
  KubeConfig,
  RbacAuthorizationV1Api,
  CoreV1Api,
  V1ConfigMap,
  V1Role,
  V1RoleBinding,
} from "@kubernetes/client-node";

export class K8sClient {
  #rbacApi: RbacAuthorizationV1Api;
  #coreApi: CoreV1Api;

  constructor() {
    const kc = new KubeConfig();
    kc.loadFromDefault();
    this.#rbacApi = kc.makeApiClient(RbacAuthorizationV1Api);
    this.#coreApi = kc.makeApiClient(CoreV1Api);
  }

  async readRole(request: Read) {
    const { name, namespace } = request;
    return unwrap(this.#rbacApi.readNamespacedRole(name, namespace));
  }

  async createRole(request: Create<V1Role>) {
    const { namespace, body } = request;
    return unwrap(this.#rbacApi.createNamespacedRole(namespace, body));
  }

  async deleteRole(request: Delete) {
    const { name, namespace } = request;
    return unwrap(this.#rbacApi.deleteNamespacedRole(name, namespace));
  }

  async readRoleBinding(request: Read) {
    const { name, namespace } = request;
    return unwrap(this.#rbacApi.readNamespacedRoleBinding(name, namespace));
  }

  async createRoleBinding(request: Create<V1RoleBinding>) {
    const { namespace, body } = request;
    return unwrap(this.#rbacApi.createNamespacedRoleBinding(namespace, body));
  }

  async deleteRoleBinding(request: Delete) {
    const { name, namespace } = request;
    return unwrap(this.#rbacApi.deleteNamespacedRoleBinding(name, namespace));
  }

  // ConfigMap
  async readConfigMap(request: Read) {
    const { name, namespace } = request;
    // Permission boundary: only allow reading the aws-auth configmap
    // if (name !== "aws-auth") {
    //   throw httpError(403, "Forbidden", JSON.stringify(new Error("Not allowed to read ConfigMap")));
    // }
    return unwrap(this.#coreApi.readNamespacedConfigMap(name, namespace));
  }

  async createConfigMap(request: Create<V1ConfigMap>) {
    const { namespace, body } = request;
    return unwrap(this.#coreApi.createNamespacedConfigMap(namespace, body));
  }

  async replaceConfigMap(request: Replace<V1ConfigMap>) {
    const { name, namespace, body } = request;
    return unwrap(
      this.#coreApi.replaceNamespacedConfigMap(name, namespace, body)
    );
  }

  async patchConfigMap(request: Patch<V1ConfigMap>) {
    const { name, namespace, body } = request;
    return unwrap(
      this.#coreApi.patchNamespacedConfigMap(name, namespace, body)
    );
  }

  async deleteConfigMap(request: Delete) {
    const { name, namespace } = request;
    return unwrap(this.#coreApi.deleteNamespacedConfigMap(name, namespace));
  }
}

/**
 * Only returns the response body from the Kubernetes API server.
 * If the request errored, return the Status object, which is, again, only the response body from the API server.
 * @param promise
 * @returns
 */
const unwrap = async <T>(promise: Promise<{ body: T }>) => {
  const makeResponse = (data: any): IndirectK8sApiResponse => {
    if (!data.response) {
      throw new Error("Malformed response");
    }
    return {
      statusCode: data.response.statusCode,
      statusMessage: data.response.statusMessage,
      body: data.body,
    };
  };

  try {
    const result = await promise;
    return makeResponse(result);
  } catch (err: any) {
    try {
      return makeResponse(err);
    } catch (err2: any) {
      throw err;
    }
  }
};
