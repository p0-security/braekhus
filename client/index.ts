import axios from "axios";
import {
  JSONRPCClient,
  JSONRPCServer,
  JSONRPCServerAndClient,
} from "json-rpc-2.0";
import { omit } from "lodash";
import { Logger } from "pino";
import WebSocket from "ws";

import { createLogger } from "../log";
import { ForwardedRequest, ForwardedResponse } from "../types";
import { deferral } from "../util/deferral";
import { Backoff } from "./backoff";
import { jwt } from "./jwks";

/**
 * Bi-directional JSON RPC client
 */
export class JsonRpcClient {
  #jsonRpcClient = deferral<JSONRPCServerAndClient>();
  #connected = deferral<void>();
  #webSocketUrl: string;
  #targetUrl: string;
  #targetHostName: string;
  #clientId: string;
  #logger: Logger;

  #backoff?: Backoff;
  #webSocket?: WebSocket;
  #retryTimeout?: NodeJS.Timeout;
  #isShutdown: boolean = false;

  constructor(
    proxyConfig: { targetUrl: string; clientId: string },
    tunnelConfig: {
      host: string;
      port: number;
      insecure?: boolean;
      backoff?: Backoff;
    }
  ) {
    this.#logger = createLogger({ name: "JsonRpcClient" });
    const { targetUrl, clientId } = proxyConfig;
    this.#targetUrl = targetUrl;
    this.#targetHostName = new URL(targetUrl).hostname;
    this.#clientId = clientId;
    const { host, port, insecure, backoff } = tunnelConfig;
    this.#webSocketUrl = `ws${!insecure ? "s" : ""}://${host}:${port}`;
    this.#backoff = backoff;
    this.#jsonRpcClient.completeWith(this.create());
    this.#jsonRpcClient.promise.catch((error: any) =>
      this.#logger.error({ error }, "Error creating JSON RPC client")
    );
  }

  async create() {
    const token = await jwt(this.#clientId);
    const clientSocket = new WebSocket(this.#webSocketUrl, {
      headers: { Authorization: `Bearer ${token}` },
    });
    const client = new JSONRPCServerAndClient(
      new JSONRPCServer(),
      new JSONRPCClient((request) => {
        try {
          clientSocket.send(JSON.stringify(request));
          return Promise.resolve();
        } catch (error) {
          return Promise.reject(error);
        }
      })
    );

    clientSocket.on("error", (error) => {
      // Do not throw error. The `on("close")` handler is called, which retries the connection.
      this.#logger.warn({ error }, "websocket error");
    });

    clientSocket.on("open", () => {
      this.#logger.info("connection opened");
      // After opening the connection, send the client ID to the server
      client
        .request("setClientId", { clientId: this.#clientId })
        .then((response) => {
          this.#logger.info({ response }, "setClientId response");
          this.#backoff?.reset();
          this.#connected.resolve();
        });
    });

    clientSocket.on("close", () => {
      this.#logger.info("connection closed");
      // Keep looking for the server after closing the connection
      if (!this.#isShutdown && this.#backoff) {
        this.#retryTimeout = setTimeout(
          () => this.create(),
          this.#backoff.next()
        );
      }
    });

    clientSocket.on("message", (data, isBinary) => {
      if (isBinary) {
        this.#logger.warn("Message in binary format is not supported");
        return;
      }
      const message = data.toString("utf-8");
      client.receiveAndSend(JSON.parse(message));
    });

    clientSocket.on("ping", () => {
      // The client automatically responds to ping messages without an implementation here
      // TODO detect broken connection based on https://github.com/websockets/ws#how-to-detect-and-close-broken-connections
      this.#logger.debug("ping");
    });

    this.#webSocket = clientSocket;

    axios.interceptors.request.use((request) => {
      this.#logger.debug({ request }, "Axios request");
      return request;
    });

    axios.interceptors.response.use((response) => {
      // Do not log response object, it's lengthy and difficult to filter out the authorization header
      this.#logger.debug(
        { response: omit(response, "request") },
        "Axios response"
      );
      return response;
    });

    client.addMethod("call", async (request: ForwardedRequest) => {
      this.#logger.debug({ request }, "forwarded request");
      // The headers are modified:
      // 1. The Content-Length header may not be accurate for the forwarded request. By removing it, axios can recalculate the correct length.
      // 2. The Host header should be switched out to the host this client is targeting.
      const response = await axios({
        baseURL: this.#targetUrl,
        url: request.path,
        method: request.method,
        headers: {
          ...omit(request.headers, "content-length"),
          host: this.#targetHostName,
        },
        params: request.params,
        data: request.data,
        validateStatus: () => true, // do not throw, we return all status codes
      });
      return {
        headers: response.headers,
        status: response.status,
        statusText: response.statusText,
        data: response.data,
      } as ForwardedResponse;
    });

    return client;
  }

  async run() {
    await this.#jsonRpcClient.promise;
  }

  async waitUntilConnected() {
    await this.#connected.promise;
  }

  shutdown() {
    this.#isShutdown = true;
    clearTimeout(this.#retryTimeout);
    this.#webSocket?.close();
  }
}
