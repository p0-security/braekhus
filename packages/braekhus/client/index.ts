import axios from "axios";
import {
  JSONRPCClient,
  JSONRPCServer,
  JSONRPCServerAndClient,
} from "json-rpc-2.0";
import { omit } from "lodash-es";
import { Logger } from "pino";
import WebSocket from "ws";

import { DEFAULT_FORWARDED_REQUEST_TIMEOUT_MILLIS } from "../common/constants.js";
import { createLogger } from "../log/index.js";
import {
  ForwardedRequest,
  ForwardedResponse,
  JQ_HEADER,
} from "../types/index.js";
import { deferral } from "../util/deferral.js";
import { Backoff } from "./backoff.js";
import { jqTransform } from "./filter.js";
import { jwt } from "./jwks.js";

const PING_TIMEOUT_MS = 30000;

/**
 * Bi-directional JSON RPC client
 */
export class JsonRpcClient {
  #jsonRpcClient = deferral<JSONRPCServerAndClient>();
  #connected = deferral<void>();
  #webSocketUrl: string;
  #targetUrl: string;
  #targetHostName: string;
  #clientId: string;
  #jwkPath: string;
  #logger: Logger;

  #backoff?: Backoff;
  #webSocket?: WebSocket;
  #pingTimeout?: NodeJS.Timeout;
  #retryTimeout?: NodeJS.Timeout;
  #isShutdown: boolean = false;

  constructor(
    proxyConfig: { targetUrl: string; clientId: string; jwkPath: string },
    tunnelConfig: {
      host: string;
      port: number;
      insecure?: boolean;
      backoff?: Backoff;
    }
  ) {
    this.#logger = createLogger({ name: "JsonRpcClient" });
    const { targetUrl, clientId, jwkPath } = proxyConfig;
    this.#targetUrl = targetUrl;
    this.#targetHostName = new URL(targetUrl).hostname;
    this.#clientId = clientId;
    this.#jwkPath = jwkPath;
    const { host, port, insecure, backoff } = tunnelConfig;
    this.#webSocketUrl = `ws${!insecure ? "s" : ""}://${host}:${port}`;
    this.#backoff = backoff;
    this.#jsonRpcClient.completeWith(this.create());
    this.#jsonRpcClient.promise.catch((error: any) =>
      this.#logger.error({ error }, "Error creating JSON RPC client")
    );

    axios.interceptors.request.use((request) => {
      this.#logger.debug({ request }, "Axios request");
      return request;
    });

    axios.interceptors.response.use((response) => {
      this.#logger.debug({ response }, "Axios response");
      return response;
    });
  }

  #scheduleReconnect() {
    // Reset the connected deferral so waitUntilConnected() works for reconnection
    if (this.#connected.isCompleted()) {
      this.#connected = deferral<void>();
    }

    if (!this.#isShutdown && this.#backoff) {
      this.#retryTimeout = setTimeout(
        () => this.create(),
        this.#backoff.next()
      );
    }
  }

  #resetPingTimeout() {
    clearTimeout(this.#pingTimeout);
    // Use `WebSocket#terminate()`, which immediately destroys the connection,
    // instead of `WebSocket#close()`, which waits for the close timer.
    // Close handler will retry connection
    this.#pingTimeout = setTimeout(() => {
      this.#logger.warn("Ping timeout - terminating connection");
      this.#webSocket?.terminate();
    }, PING_TIMEOUT_MS);
  }
  async create() {
    let token: string;
    let clientSocket: WebSocket;

    try {
      token = await jwt(this.#jwkPath, this.#clientId);
      clientSocket = new WebSocket(this.#webSocketUrl, {
        headers: { Authorization: `Bearer ${token}` },
      });
    } catch (error) {
      this.#logger.error({ error }, "Failed to create WebSocket connection");
      this.#scheduleReconnect();
      throw error;
    }

    const client = new JSONRPCServerAndClient(
      new JSONRPCServer(),
      new JSONRPCClient((request) => {
        try {
          clientSocket.send(JSON.stringify(request));
          return Promise.resolve();
        } catch (error) {
          return Promise.reject(error);
        }
      })
    );

    clientSocket.on("error", (error) => {
      // Do not throw error. The `on("close")` handler is called, which retries the connection.
      this.#logger.warn({ error }, "websocket error");
    });

    clientSocket.on("open", () => {
      this.#logger.info("connection opened");
      // Start ping timeout
      this.#resetPingTimeout();
      // After opening the connection, send the client ID to the server
      client
        .request("setClientId", { clientId: this.#clientId })
        .then((response) => {
          this.#logger.info({ response }, "setClientId response");
          this.#backoff?.reset();
          this.#connected.resolve();
        });
    });

    clientSocket.on("close", () => {
      this.#logger.info("connection closed");
      // Keep looking for the server after closing the connection
      this.#scheduleReconnect();
    });

    clientSocket.on("message", (data, isBinary) => {
      if (isBinary) {
        this.#logger.warn("Message in binary format is not supported");
        return;
      }
      const message = data.toString("utf-8");
      client.receiveAndSend(JSON.parse(message));
    });

    clientSocket.on("ping", () => {
      // The client automatically responds to ping messages without an implementation here
      this.#logger.debug("ping");
      this.#resetPingTimeout();
    });

    this.#webSocket = clientSocket;

    client.addMethod("call", async (request: ForwardedRequest) => {
      this.#logger.debug(
        {
          request: {
            ...request,
            headers: omit(request.headers, "authorization"),
          },
        },
        "forwarded request"
      );
      // The headers are modified:
      // 1. The Content-Length header may not be accurate for the forwarded request. By removing it, axios can recalculate the correct length.
      // 2. The Host header should be switched out to the host this client is targeting.
      const response = await axios({
        baseURL: this.#targetUrl,
        url: request.path,
        method: request.method,
        headers: {
          ...omit(request.headers, "content-length"),
          host: this.#targetHostName,
        },
        params: request.params,
        data: request.data,
        validateStatus: () => true, // do not throw, we return all status codes
        timeout:
          request.options?.timeoutMillis ||
          DEFAULT_FORWARDED_REQUEST_TIMEOUT_MILLIS,
      });
      const jpSelectHeader = request.headers[JQ_HEADER];
      this.#logger.debug({ response }, "forwarded response before filters");
      const data = await jqTransform(response.data, jpSelectHeader);
      this.#logger.debug(
        { response: data },
        "forwarded response data after filters"
      );
      return {
        headers: response.headers,
        status: response.status,
        statusText: response.statusText,
        data,
      } as ForwardedResponse;
    });

    return client;
  }

  async run() {
    await this.#jsonRpcClient.promise;
  }

  async waitUntilConnected() {
    await this.#connected.promise;
  }

  shutdown() {
    this.#isShutdown = true;
    clearTimeout(this.#retryTimeout);
    this.#webSocket?.close();
  }
}
