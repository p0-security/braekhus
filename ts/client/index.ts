import axios from "axios";
import {
  JSONRPCClient,
  JSONRPCServer,
  JSONRPCServerAndClient,
} from "json-rpc-2.0";
import { omit } from "lodash";
import { Logger } from "pino";
import { ForwardedRequest, ForwardedResponse } from "types";
import WebSocket from "ws";

import { createLogger } from "../log";
import { deferral } from "../util/deferral";
import { jwt } from "./jwks";

const CONNECT_RETRY_INTERVAL_MILLIS = 3000;
const KEEP_ALIVE_INTERVAL_MILLIS = 5000;

/**
 * Bi-directional JSON RPC client
 */
export class JsonRpcClient {
  #jsonRpcClient = deferral<JSONRPCServerAndClient>();
  #connected = deferral<void>();
  #webSocketUrl: string;
  #httpUrl: string;
  #targetUrl: string;
  #targetHostName: string;
  #clientId: string;
  #logger: Logger;

  #webSocket?: WebSocket;
  #keepAliveTimeout?: NodeJS.Timeout;
  #retryTimeout?: NodeJS.Timeout;
  #isShutdown: boolean = false;

  constructor(
    proxyConfig: { targetUrl: string; clientId: string },
    tunnelConfig: { host: string; port: number; insecure?: boolean }
  ) {
    this.#logger = createLogger({ name: "JsonRpcClient" });
    const { targetUrl, clientId } = proxyConfig;
    this.#targetUrl = targetUrl;
    this.#targetHostName = new URL(targetUrl).hostname;
    this.#clientId = clientId;
    const { host, port, insecure } = tunnelConfig;
    this.#webSocketUrl = `ws${!insecure ? "s" : ""}://${host}:${port}`;
    this.#httpUrl = `http${!insecure ? "s" : ""}://${host}:${port}`;
    this.#jsonRpcClient.completeWith(this.create());
    this.#jsonRpcClient.promise.catch((error: any) =>
      this.#logger.error({ error }, "Error creating JSON RPC client")
    );
  }

  async create() {
    const token = await jwt();
    const clientSocket = new WebSocket(this.#webSocketUrl, {
      headers: { Authorization: `Bearer ${token}` },
    });
    const client = new JSONRPCServerAndClient(
      new JSONRPCServer(),
      new JSONRPCClient((request) => {
        try {
          clientSocket.send(JSON.stringify(request));
          return Promise.resolve();
        } catch (error) {
          return Promise.reject(error);
        }
      })
    );

    const keepAlive = () => {
      axios
        .request({
          baseURL: this.#httpUrl,
          url: "/live",
          method: "GET",
          validateStatus: () => true, // do not throw, we return all status codes
        })
        .then((response) => {
          this.#logger.debug({ response }, "live response");
          this.#keepAliveTimeout = setTimeout(
            keepAlive,
            KEEP_ALIVE_INTERVAL_MILLIS
          );
        });
      // client.request("live", { clientId: this.#clientId }).then((response) => {
      //   this.#logger.debug({ response }, "live response");
      //   this.#keepAliveTimeout = setTimeout(
      //     keepAlive,
      //     KEEP_ALIVE_INTERVAL_MILLIS
      //   );
      // });
    };

    clientSocket.on("error", (error) => {
      // Do not throw error. The `on("close")` handler is called, which retries the connection.
      this.#logger.warn({ error }, "websocket error");
    });

    clientSocket.on("open", () => {
      this.#logger.info("connection opened");
      // After opening the connection, send the client ID to the server
      client
        .request("setClientId", { clientId: this.#clientId })
        .then((response) => {
          this.#logger.info({ response }, "setClientId response");
          this.#connected.resolve();
        });
      this.#keepAliveTimeout = setTimeout(
        keepAlive,
        KEEP_ALIVE_INTERVAL_MILLIS
      );
    });

    clientSocket.on("close", () => {
      this.#logger.info("connection closed");
      // Keep looking for the server after closing the connection
      if (!this.#isShutdown) {
        this.#retryTimeout = setTimeout(
          () => this.create(),
          CONNECT_RETRY_INTERVAL_MILLIS
        );
      }
    });

    clientSocket.on("message", (data, isBinary) => {
      if (isBinary) {
        this.#logger.warn("Message in binary format is not supported");
        return;
      }
      const message = data.toString("utf-8");
      client.receiveAndSend(JSON.parse(message));
    });

    clientSocket.on("ping", () => {
      this.#logger.debug("ping");
    });

    this.#webSocket = clientSocket;

    axios.interceptors.request.use((request) => {
      this.#logger.debug({ request }, "Axios request");
      return request;
    });

    axios.interceptors.response.use((response) => {
      this.#logger.debug({ response }, "Axios response");
      return response;
    });

    client.addMethod("live", ({}) => {
      return { ok: true };
    });
    client.addMethod("call", async (request: ForwardedRequest) => {
      this.#logger.info({ request }, "forwarded request");
      // The headers are modified:
      // 1. The Content-Length header may not be accurate for the forwarded request. By removing it, axios can recalculate the correct length.
      // 2. The Host header should be switched out to the host this client is targeting.
      const response = await axios({
        baseURL: this.#targetUrl,
        url: request.path,
        method: request.method,
        headers: {
          ...omit(request.headers, "content-length"),
          host: this.#targetHostName,
        },
        params: request.params,
        data: request.data,
        validateStatus: () => true, // do not throw, we return all status codes
      });
      return {
        headers: response.headers,
        status: response.status,
        statusText: response.statusText,
        data: response.data,
      } as ForwardedResponse;
    });

    return client;
  }

  async run() {
    await this.#jsonRpcClient.promise;
  }

  async waitUntilConnected() {
    await this.#connected.promise;
  }

  shutdown() {
    this.#isShutdown = true;
    clearTimeout(this.#keepAliveTimeout);
    clearTimeout(this.#retryTimeout);
    this.#webSocket?.close();
  }
}
