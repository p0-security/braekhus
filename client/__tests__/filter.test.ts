import { jpFilter as jqFilter } from "client/filter";

const exampleResponse = {
  headers: {
    "audit-id": "6a816d03-d29f-4657-b624-35766c8a5572",
    "cache-control": "no-cache, private",
    "content-type": "application/json",
    "x-kubernetes-pf-flowschema-uid": "ef7c3d21-4649-422d-a4fa-e5b85d52f141",
    "x-kubernetes-pf-prioritylevel-uid": "a6704cd4-c4fb-48bb-982e-a1ed79bfb12d",
    date: "Wed, 25 Oct 2023 06:08:30 GMT",
    "transfer-encoding": "chunked",
  },
  status: 200,
  statusText: "OK",
  data: {
    kind: "ClusterRoleList",
    apiVersion: "rbac.authorization.k8s.io/v1",
    metadata: {
      resourceVersion: "4423772",
    },
    items: [
      {
        metadata: {
          name: "admin",
          uid: "4251609f-b3ca-4875-8ec6-76b40ab62748",
          resourceVersion: "345",
          creationTimestamp: "2023-10-02T22:51:23Z",
          labels: {
            "kubernetes.io/bootstrapping": "rbac-defaults",
          },
          annotations: {
            "rbac.authorization.kubernetes.io/autoupdate": "true",
          },
          managedFields: [
            {
              manager: "clusterrole-aggregation-controller",
              operation: "Apply",
              apiVersion: "rbac.authorization.k8s.io/v1",
              time: "2023-10-02T22:51:38Z",
              fieldsType: "FieldsV1",
              fieldsV1: {
                "f:rules": {},
              },
            },
            {
              manager: "kube-apiserver",
              operation: "Update",
              apiVersion: "rbac.authorization.k8s.io/v1",
              time: "2023-10-02T22:51:23Z",
              fieldsType: "FieldsV1",
              fieldsV1: {
                "f:aggregationRule": {
                  ".": {},
                  "f:clusterRoleSelectors": {},
                },
                "f:metadata": {
                  "f:annotations": {
                    ".": {},
                    "f:rbac.authorization.kubernetes.io/autoupdate": {},
                  },
                  "f:labels": {
                    ".": {},
                    "f:kubernetes.io/bootstrapping": {},
                  },
                },
              },
            },
          ],
        },
      },
      {
        metadata: {
          name: "aws-node",
          uid: "63a3bed8-a750-440d-bca9-b5f41c8786c2",
          resourceVersion: "273",
          creationTimestamp: "2023-10-02T22:51:35Z",
          labels: {
            "app.kubernetes.io/instance": "aws-vpc-cni",
            "app.kubernetes.io/name": "aws-node",
            "app.kubernetes.io/version": "v1.12.6",
            "k8s-app": "aws-node",
          },
          annotations: {
            "kubectl.kubernetes.io/last-applied-configuration":
              '{"apiVersion":"rbac.authorization.k8s.io/v1","kind":"ClusterRole","metadata":{"annotations":{},"labels":{"app.kubernetes.io/instance":"aws-vpc-cni","app.kubernetes.io/name":"aws-node","app.kubernetes.io/version":"v1.12.6","k8s-app":"aws-node"},"name":"aws-node"},"rules":[{"apiGroups":["crd.k8s.amazonaws.com"],"resources":["eniconfigs"],"verbs":["list","watch","get"]},{"apiGroups":[""],"resources":["namespaces"],"verbs":["list","watch","get"]},{"apiGroups":[""],"resources":["pods"],"verbs":["list","watch","get"]},{"apiGroups":[""],"resources":["nodes"],"verbs":["list","watch","get","update"]},{"apiGroups":["extensions"],"resources":["*"],"verbs":["list","watch"]},{"apiGroups":["","events.k8s.io"],"resources":["events"],"verbs":["create","patch","list"]}]}\n',
          },
          managedFields: [
            {
              manager: "kubectl-client-side-apply",
              operation: "Update",
              apiVersion: "rbac.authorization.k8s.io/v1",
              time: "2023-10-02T22:51:35Z",
              fieldsType: "FieldsV1",
              fieldsV1: {
                "f:metadata": {
                  "f:annotations": {
                    ".": {},
                    "f:kubectl.kubernetes.io/last-applied-configuration": {},
                  },
                  "f:labels": {
                    ".": {},
                    "f:app.kubernetes.io/instance": {},
                    "f:app.kubernetes.io/name": {},
                    "f:app.kubernetes.io/version": {},
                    "f:k8s-app": {},
                  },
                },
                "f:rules": {},
              },
            },
          ],
        },
      },
    ],
  },
};

describe("json path filter", () => {
  it("returns only selected fields", async () => {
    const result = await jqFilter(
      exampleResponse,
      "{ headers: .headers, status: .status, statusText: .statusText, data: { kind: .data.kind, items: [.data.items[] | { metadata: { name: .metadata.name } }] } }"
    );

    expect(result).toMatchSnapshot();
  });
});
