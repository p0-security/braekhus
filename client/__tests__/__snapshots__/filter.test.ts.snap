// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`json path filter returns only selected fields 1`] = `
{
  "data": {
    "items": [
      {
        "metadata": {
          "name": "admin",
        },
      },
      {
        "metadata": {
          "name": "aws-node",
        },
      },
    ],
    "kind": "ClusterRoleList",
  },
  "headers": {
    "audit-id": "6a816d03-d29f-4657-b624-35766c8a5572",
    "cache-control": "no-cache, private",
    "content-type": "application/json",
    "date": "Wed, 25 Oct 2023 06:08:30 GMT",
    "transfer-encoding": "chunked",
    "x-kubernetes-pf-flowschema-uid": "ef7c3d21-4649-422d-a4fa-e5b85d52f141",
    "x-kubernetes-pf-prioritylevel-uid": "a6704cd4-c4fb-48bb-982e-a1ed79bfb12d",
  },
  "status": 200,
  "statusText": "OK",
}
`;
